### PR TITLE
Scanner Peep.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
[Compiler VB Scanner] Implement Peep and usage within the scanner.
* ScannerXML
* Scanner
* ScannerInterpolatedString

`Peep` is a integration of the contents of `CanGet` and `Peek` as these often appear in close proximity.  By combining them the source code is shorter, and potentially slightly quicker as the 'ch` value can be cached, and thus reused later. Reducing repetition of multiple function calls.
(On my machine time to run the tests reduced from 21mins to 20mins) 

Note: `Peep` is similar to a `TryParse` function but it implements different semantics in the `false` case, where it preserves the existing value contain in `ch`

There may be some potential reduction in the number of temporary char variables, with closer examination, but this can be left for another PR.

### CODE REVIEW

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
I'm doing a few refinements of the temporary char variables I've introduced, see which I remove (reuse an existing char variable.

**Comment by davkean** [comment]
tag @dotnet/roslyn-compiler

**Comment by AdamSpeight2008** [comment]
Test time down to around 18-19 minutes. (On my machine)

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
Ready for review.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 1]
```diff
@@ -1033,7 +1021,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                     Return If(dl, MakeHashToken(precedingTrivia, fullWidth))
 
                 Case "&"c
-                    If CanGet(1) AndAlso BeginsBaseLiteral(Peek(1)) Then
+                    If Peep(1, ch) AndAlso BeginsBaseLiteral(ch) Then
```
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Why is this one using `ch` and the others us `cx`?

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 2]
```diff
@@ -1237,16 +1225,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
         End Function
 
         Private Function ScanRightAngleBracket(precedingTrivia As SyntaxList(Of VisualBasicSyntaxNode), charIsFullWidth As Boolean) As SyntaxToken
-            Debug.Assert(CanGet)  ' >
-            Debug.Assert(Peek() = ">"c OrElse Peek() = FULLWIDTH_GREATER_THAN_SIGN)
-
+            Dim c As Char
+#If DEBUG Then
```
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Did you mean to include this debug block?

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 3]
```diff
@@ -1769,17 +1731,17 @@ FullWidthRepeat2:
                         End If
 
                     Case "U"c, "u"c
-                        If literalKind <> NumericLiteralKind.Float AndAlso CanGet(Here + 1) Then
-                            Dim NextChar As Char = Peek(Here + 1)
+                        If literalKind <> NumericLiteralKind.Float AndAlso Peep(Here + 1, ch) Then
+                            ' Dim ch As Char = Peek(Here + 1)
```
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Remove comment

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 4]
```diff
@@ -2000,7 +1953,7 @@ FullWidthRepeat2:
             Dim YearIsTwoDigits As Boolean = False
             Dim DaysToMonth As Integer() = Nothing
             Dim yearIsFirst As Boolean = False
-
+            Dim ch, cx As Char, sep1 As Char
```
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Having ch, cx, and sep1 used heavily hurts readability.  Can these have better names?

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 3 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Will do.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 1 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: @TyOverby 
I'll re-examine this method.
Some of the case can alter the value of `ch` as all code path in that `case` exit from the function. It is the `case`'s that don't I have to be careful with as `ch` is used after the `Select` block.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 4 above)
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: @TyOverby 
`ch` or `c` is commonly used throughout the scanner code to mean **this character**.
`cx` I'm using to mean some other character. It is typical used where a change to the value of `ch` would affects it meaning. Typically in the `ElseIf` condition.

`sep1` is being renamed to `FirstDateSeparator`

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot retest this please.

**Comment by AdamSpeight2008** [comment]
Re-examined the usage of `Peep` and reduced the number of temporary `char` variables.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Anybody care to comment on this? @AlekseyTs ?

**Comment by AdamSpeight2008** [comment]
@AlekseyTs I've also got a branch that integrate this into the `future` branch.

**Comment by TyOverby** [comment]
I went through this a while ago and it all looked good.  I'll review the stuff that has happened since then.

**Comment by jaredpar** [comment]
> By combining them the source code is shorter, and potentially slightly quicker as the 'ch` value can be cached, and thus reused later. Reducing repetition of multiple function calls.
(On my machine time to run the tests reduced from 21mins to 20mins)

Can you publish the tests?  Perf claims must be backed up with tests.

**Comment by TyOverby** [comment]
@AdamSpeight2008 I think the last few "remove temporaries" commits are going a bit to far.

```
             If IsLeftCurlyBracket(c) Then
                Return Not Peep(offset + 1, c) OrElse Not IsLeftCurlyBracket(c)
```

I think this code is more unreadable compared to the version that used temporaries.

Compilers and jitting engines are typically pretty good at dealing with temporaries that outlive their use.

**Comment by AdamSpeight2008** [comment]
@jaredpar 
The "perf" improvements are base on the times of the build and test. Current Master vs PR
`msbuild /v:m /m BuildAndTest.proj /p:PublicBuild=true /p:DeployExtension=false`
It just a first approximation guide, if I see >1min improvement, then that is highly unlikey due random jitter.

-------------------------
@TyOverby 
I think that code is "unreadable" due the to negatation on the inputs.
Could be rewritten as.
```vb
If IsLeftCurlyBracket(c) Then
  Return Not ( Peep(offset + 1, c) AndAlso  IsLeftCurlyBracket(c))
```

Reusing the `select case`'s *selection variable* is valid.
* As long as all paths within the `case selection` block return or exit the function.
* If it doesn't maybe possible to use, but a temporary is suggested.

This is because the choice of which `case selection block` to use, happens at the `Select Case foo`.
It the compiler rewrites it into some other form, it as had keep the same semantics.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 5]
```diff
@@ -1376,8 +1365,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             ' // The C++ compiler refuses to inline IsIdentifierCharacter, so the
             ' // < 128 test is inline here. (This loop gets a *lot* of traffic.)
             ' TODO: make sure we get good perf here
-            While CanGet(len)
-                ch = Peek(len)
+ 
+            While Peep(len, ch)
```
Comment (AdamSpeight2008) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: I have a version of this `while` loop that becomes.
```vb
' // The C++ compiler refuses to inline IsIdentifierCharacter, so the
' // < 128 test is inline here. (This loop gets a *lot* of traffic.)
' TODO: make sure we get good perf here
While Peep(len, ch) AndAlso (IsNarrowIdentifierCharacter(Convert.ToUInt16(ch)) OrElse IsWideIdentifierCharacter(ch))
  len += 1
End While
```
by push one of the condition into the two called functions.
```
        Friend Shared Function IsNarrowIdentifierCharacter(c As UInt16) As Boolean
            Return c < 128US AndAlso s_isIDChar(c)
        End Function

        ''' <summary>
        ''' Determines if the Unicode character can be a part of a Visual Basic identifier.
        ''' </summary>
        ''' <param name="c">The Unicode character.</param>
        ''' <returns>A boolean value set to True if character can be part of a valid identifier.</returns>
        Public Shared Function IsIdentifierPartCharacter(c As Char) As Boolean
            If c < ChrW(128) Then
                Return IsNarrowIdentifierCharacter(Convert.ToUInt16(c))
            End If

            Return IsWideIdentifierCharacter(c)
        End Function
```

**Comment by jaredpar** [comment]
> It just a first approximation guide, if I see >1min improvement, then that is highly unlikey due random jitter.

Disagree.  I regularly see times fluctuate at that level.  Build and test incorporates many elements and minute level fluctuations are expected.   

To test the perf of a parser level change, particularly the scanner, I would be looking for actual direct tests of the scanner, run across several iterations to help isolate out noise.

**Comment by davkean** [comment]
@AdamSpeight2008 Are you planning on updating the PR with direct tests of the scanner?

**Comment by gafter** [comment]
I think it would be fine to have performance tests that are not checked in to validate that the proposed change actually improves things. It would not be obvious how to turn such a performance test into a unit test suitable for the regression suite.

Having said that, I do believe it is important to demonstrate the effect of this PR on the behavior of the scanner before we accept the changes.

**Comment by AdamSpeight2008** [comment]
@davkean @gafter I've written a perf reporter but when I run it against the roslyn solution I get an NRE (see #7206).

**Comment by AdamSpeight2008** [comment]
Used a different way.
(https://gist.github.com/AdamSpeight2008/6f28d729d522271c0a50)

**Comment by AdamSpeight2008** [comment]
Gist of Timings.
https://gist.github.com/AdamSpeight2008/6f28d729d522271c0a50

**Comment by AdamSpeight2008** [comment]
Total time is slight regression: `-0.0052s", on a file level it is ~82% better.
```
  Improved: 2208
      Same: 268
Regression: 547
```

**Comment by gafter** [comment]
@AdamSpeight2008 I think your perf measurements are dominated by a start-up transient that you are accounting for as due to your changes. I suggest scanning a few files (with both old and new scanner) and throwing away the first few results before starting to record the results for comparison.

**Comment by AdamSpeight2008** [comment]
@gafter Re-done them.
**Test Code**
```vb
    Sub Main(args() As String)
        Dim vbfiles = IO.Directory.EnumerateFiles(args(0), "*.vb", IO.SearchOption.AllDirectories)
        For i = 0 To 9
            Dim sn = SyntaxFactory.ParseSyntaxTree(vbfiles(0))
        Next
        Dim res = <results total=""></results>
        Dim total = TimeSpan.Zero
        Dim sw As New System.Diagnostics.Stopwatch
        For Each vbfile In vbfiles
            GC.Collect()

            sw.Restart()
            Dim sn = SyntaxFactory.ParseSyntaxTree("",, vbfile)
            sw.Stop()
            total = total.Add(sw.Elapsed)
            res.Add(<file timing=<%= sw.Elapsed.ToString %> name=<%= vbfile %>/>)
        Next
        res.@total = total.ToString
        res.Save(args(1))
        Console.WriteLine()
        Console.WriteLine(total.ToString)
    End Sub
```
All I change is the DLLs (original vs proposal)
[Results Gist](https://gist.github.com/AdamSpeight2008/2db026705aeb2845e10d)

~5ms improvement. ~90% of the files are the same or better.

**Comment by AdamSpeight2008** [comment]
I wouldn't say my PC is "standard". `2 x X5960 (6 cpu cores + Hyper-Threaded ) with 40Gb of DDR3 ECC RAM``` There only running at about 1.7GHz but I can go up close to 4GHz.

**Comment by gafter** [comment]
You appear to be using this method:

```vb
        ''' <summary>
        ''' Produces a syntax tree by parsing the source text.
        ''' </summary>
        Public Shared Function ParseSyntaxTree(
            text As String,
            Optional options As ParseOptions = Nothing,
            Optional path As String = "",
            Optional encoding As Encoding = Nothing,
            Optional cancellationToken As CancellationToken = Nothing) As SyntaxTree

            Return ParseSyntaxTree(SourceText.From(text, encoding), options, path, cancellationToken)
        End Function
```

In which case you're just parsing the empty string over and over again.

**Comment by gafter** [comment]
The cost of handling garbage appears to be ignored by your test. Please move the timing outside the loop (we only need one time for parsing everything), and remove `GC.Collect()` from the code. Also, I think you need to change what API you're calling so that code actually gets parsed.

**Comment by AdamSpeight2008** [comment]
[Gist for updated results](https://gist.github.com/AdamSpeight2008/3f137aa965c577db2b26)

**Comment by AdamSpeight2008** [comment]
Slightly quicker (~.5sec), less GC Time (~20sec)

**Comment by gafter** [comment]
I'm not sure we're really that interested in how quickly the VB scanner can scan C# source files. Can you please remove all the stuff having to do with garbage-collection, and just scan all the VB source files (only) in a single timing measurement? You also seem to have lost the code for warming up the test before beginning measurement. A simple way of doing that would be to exclude the first few files from the timing measurements and start the stopwatch after, say, the third source file.

**Comment by AdamSpeight2008** [comment]
[Gist of timing results](https://gist.github.com/AdamSpeight2008/9e959461e79a7033951a)
@gafter A single total timing has been included in every one, look at the total attributes on the `<result>` node.
original 
```xml
<results total0="00:00:03.9249847" total1="00:02:21.7224061">
```
proposal
```xml
<results total0="00:00:03.8988103" total1="00:02:20.5710116">
```

`total0` is the total parse time (without GC Collection)
`total1` is the total parse time (if you include the GC Collect afterwards)
on a doc level
 `timing` is the parse time (without GC Collection) 
`postGC` is the parse time (if the GC Collect is included).

**Comment by gafter** [comment]
@AdamSpeight2008 Your "single total time" is a sum of lots of little snippets of timing. I'm asking you to start and stop the clock only once, and to run GC at most once at the end.

**Comment by AdamSpeight2008** [comment]
The stopping and starting of the stopwatch, is to remove the code that does the test, displays output from the timings. By having document level timings, will make it possible to deeper analysis.

[Zip of Results and Code](http://s000.tinyupload.com/index.php?file_id=95773912557476773278)
```
Original: <results total0="00:06:06.7785601" total1="00:13:12.6519269">
Proposal: <results total0="00:06:05.4573658" total1="00:13:27.1300621">
```
More in depth analysis of the results, suggest that the GC results (`total1`) is affected by a small percentage of parsings.
```
Parse
 Better: 395425
  Equal:    174  = 395599 66.33% 
  Worse: 200778  = 596377
Parse + GC
 Better: 515600  
  Equal: 0       = 515600 86.46%
  Worse: 80777   = 596377
```

**Comment by gafter** [comment]
Your zip file does not have your test program, so I cannot see what it is doing and I have no idea what you're measuring now. I don't know what kind of "deeper analysis" you have planned, as you have not told us anything about it. If you remove the code that does timings on a per-file basis as I suggested, you will not have to stop and start the stopwatch to exclude that code from the timings. We don't need thousands of measurements.

Since we are not interested in the behavior of the CLR when it is asked to collect garbage far more frequently than it would normally do, any test that forces frequent GC does not provide useful measurements.

**Comment by AdamSpeight2008** [comment]
@gafter  The test program is included, the imaginatively titled `ConsoleApplication1` in each of the 
folders. *Just decompile it with your .net decompiler of choice.*
```
Option Infer On

Module Module1

    Sub Main(args() As String)
        Dim SLN = args(0)
        Dim ws = Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.Create()
        Dim total0 = TimeSpan.Zero
        Dim total1 = TimeSpan.Zero
        Dim res = <results total0="" total1=""></results>
        Using ws
            ws.SkipUnrecognizedProjects = True
            Console.Write("Loading Solution [")
            Dim tsk = ws.OpenSolutionAsync(SLN)
            Do
                Task.Delay(1000).Wait()
                Console.Write(".")
            Loop Until tsk.IsCompleted
            Console.WriteLine("]")
            Console.WriteLine()
            Dim soln = tsk.Result
            Dim sw As New System.Diagnostics.Stopwatch
            For Each p In soln.Projects.Where(Function(p0) p0.Language = Microsoft.CodeAnalysis.LanguageNames.VisualBasic)
                '
                '  Go through each project in the solution 
                '
                Dim pn As Xml.Linq.XElement = <Project name=<%= p.Name %> path=<%= p.OutputFilePath %>></Project>
                Console.WriteLine($"{p.Name} ")

                For Each d In p.Documents.Where(Function(__) __.SourceCodeKind = Microsoft.CodeAnalysis.SourceCodeKind.Regular)
                    '
                    ' Go through each document in solution.
                    '
                    Dim sourceCode = d.GetTextAsync.Result
                    Dim node As Microsoft.CodeAnalysis.SyntaxTree = Nothing
                    ' Warmup the parser logic
                    Write("Pre-Warm [", ConsoleColor.Red)
                    For run = 0 To 3
                        node = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseSyntaxTree(sourceCode)
                        Write(".", ConsoleColor.Red)
                    Next
                    Write("]  ", ConsoleColor.Red)
                    '
                    ' Clear the results of the timings
                    '
                    Dim pre_GC = TimeSpan.Zero
                    Dim post_Gc = TimeSpan.Zero
                    ' Reset the stopwatch
                    sw.Reset()
                    Write("Run [", ConsoleColor.Blue)
                    '
                    ' Parse the document 100 times.
                    '
                    For run = 1 To 100
                        '
                        ' start the timing of the parse
                        '
                        sw.Start()
                        node = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseSyntaxTree(sourceCode)
                        sw.Stop()
                        '
                        ' STOP the timing
                        '
                        If run Mod 10 = 0 Then Write(".", ConsoleColor.Blue)
                    Next
                    Write("]  ", ConsoleColor.Blue)
                    pre_GC = sw.Elapsed
                    '
                    '  CONTINUE the timing for the GC Collection times. ( If it produces less garbage, it take less time)
                    '
                    sw.Start()
                    GC.Collect()
                    sw.Stop()
                    '
                    ' STOP the stop watch
                    '
                    post_Gc = sw.Elapsed
                    '
                    '  Record the results
                    '
                    Dim tn = <doc timing=<%= pre_GC.ToString() %> postGC=<%= post_Gc.ToString() %> name=<%= d.Name %>/>
                    pn.Add(tn)
                    WriteLine($"  {pre_GC.ToString} {post_Gc.ToString}  {d.Name}", ConsoleColor.White)
                    '
                    '  Acculmalate the timing.
                    '
                    total0 = total0.Add(pre_GC)
                    total1 = total1.Add(post_Gc)
                    res.Add(pn)
                Next
                Console.WriteLine()

            Next
        End Using
        '
        '  ADD Acculmalation Timings into results
        '
        res.@total0 = total0.ToString
        res.@total1 = total1.ToString
        '
        '  SAVE Results
        '
        res.Save(args(1))
        '
        ' Show Acculmated timings
        '
        Console.WriteLine()
        Console.WriteLine($"Total: {total0.ToString} + GC{total1.ToString}")

    End Sub


    Private Sub Write(text As String, inCol As ConsoleColor)
        Dim orig = Console.ForegroundColor
        Console.ForegroundColor = inCol
        Console.Write(text)
        Console.ForegroundColor = orig
    End Sub

    Private Sub WriteLine(text As String, inCol As ConsoleColor)
        Dim orig = Console.ForegroundColor
        Console.ForegroundColor = inCol
        Console.WriteLine(text)
        Console.ForegroundColor = orig
    End Sub
End Module

```

If I just set the stopwatch running at the start and stop at the end the program, the timings will be meaningless as it'll include stuff, that is nothing to do with the parse times. Starting just the stopwatch, just point the parse takes place, and stopping it afterwards. Better reflects what we are testing. 

Separation of the parse time and the time it takes to tidy up afterwards. Is useful information as if the "improvement" produces lot of objects, it will take more time for the GC to run.
It also levels the playing field for each parse time. less likely to hit a GC collection during it.

File Level times would allow us (me) to know what percentage of file show improvement / worsen and importantly which. 
* Is it just a small percentage of files causing a bias? 
* Are the worse performing files related someway?
* Is worse time minor?

By lumping it down a single overall time, you wouldn't know. file show improvements, or worsen times. Eg Is it just a small percentage of files causing a bias? 

Say that this "test" was part of the build process, it could be use to show trends and averages. As well being used as a "standard" base, to see if Pull Request affect the parse times, or have regressed.

**Comment by AdamSpeight2008** [comment]
2nd run.
```
Original: <results total0="00:06:17.0384866" total1="00:13:42.7943763">
Proposed: <results total0="00:06:07.6998374" total1="00:13:17.5918040">
```
```
Parse
 Better: 455154
  Equal: 0
  Worse: 141223
Parse (GC)
 Better: 539248
  Equal: 0
  Worse: 57129
```

**Comment by gafter** [comment]
Your test program intermixes "warm-up" with the test run. The warm-up should happen once, before the test starts. To avoid mixing up stuff that has nothing to do with the timing, and the stuff that should be timed, *remove all the code that has nothing to do with the timing*. Remove any GC invocations that are inside any loops, as that creates an unrealistic running environment. Your conjecture that more garbage = more expensive GC is not necessarily correct, and in any case irrelevant to this timing test. You should do at most one GC at the end, inside the stopwatch.

I appreciate your intuition that more data = better, but that is not helpful to us in evaluating your PR. By all means maintain your own benchmarks for your own purposes, but please do not give it to us instead of the timing tests we need to evaluate the PR.

**Comment by AdamSpeight2008** [comment]
The test code is the same for both, so doesn't make a difference. All you require is a relative timing and I've provided that on multiple occasions. 
* Build and Test times on both my machine and your build servers.
* Parsing just the files in the solution
* Parsing the documents from the projects in the solution.
All of them have shown improvement over the existing code. 

@gafter Why don't you provide the test code then, rather than just talking about it.

**Comment by gafter** [comment]
@AdamSpeight2008 I don't provide the test code because I'm not trying to get this PR into Roslyn.

Closing due to lack of evidence on performance as described. I'll keep monitoring this PR in case you change your mind about testing.

**Comment by AdamSpeight2008** [comment]
Is this the way? @gafter 
```
  Sub Main(args() As String)
        Dim PathToSolutionFile = args(0)
        Dim Workspace = Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.Create()
        Workspace.SkipUnrecognizedProjects = True
        Dim _Solution = Workspace.OpenSolutionAsync(PathToSolutionFile).Result
        Using Workspace
            Dim docs = From _Project In _Solution.Projects
                       Where _Project.Language = Microsoft.CodeAnalysis.LanguageNames.VisualBasic
                       From _Document In _Project.Documents
                       Where _Document.SourceCodeKind = Microsoft.CodeAnalysis.SourceCodeKind.Regular
                       Select _Document

            Dim tree As Microsoft.CodeAnalysis.SyntaxTree
            For x = 0 To 9
                Dim sourceCode = docs(x).GetTextAsync.Result
                tree = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseSyntaxTree(sourceCode)
            Next
            Dim sw As New Diagnostics.Stopwatch
            sw.Start()
            For x = 1 To 100
                For Each doc In docs
                    Dim sourceCode = doc.GetTextAsync.Result
                    tree = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseSyntaxTree(sourceCode)
                Next
                GC.Collect()
            Next
            sw.Stop()
            Console.WriteLine(sw.Elapsed.ToString)
        End Using
    End Sub
```

**Comment by gafter** [comment]
@AdamSpeight2008 Move `GC.Collect()` out of the loop and make `docs` an array, and that looks great.

**Comment by AdamSpeight2008** [comment]
@gafter Ok, I'll do that and rerun the tests. Maybe a couple of hours before the results are posted.

**Comment by AdamSpeight2008** [comment]
@gafter 
Here are the results of the testing.
```
Original

Run #1  00:11:38.8527845
Run #2  00:14:07.7066477
Run #3  00:12:34.9365867
        ----------------
        00:38:21.4960189 sum
        ----------------
        00:12:40.4986726 avg


Proposed

Run #1  00:11:22.0387645
Run #2  00:13:49.6298996
Run #3  00:11:57.7907524
        ----------------
        00:36:29.4694165 sum
        ----------------
        00:12:09.8231388 avg
```

[Zipped Files](http://s000.tinyupload.com/index.php?file_id=13879352803320828001)

**Comment by gafter** [comment]
This looks good to me. @dotnet/roslyn-compiler Can I please have another set of eyes on this?

**Comment by TyOverby** [comment]
Someone at lunch mentioned that having "peek" and "peep" intermingled could confuse readers.  Fortunately we have a naming convention for functions that return bool for success/failure and use out parameters.  

What about changing "peep" to "tryPeek"?

**Comment by AdamSpeight2008** [comment]
@TyOverby  I've delibratly not used `Try` as it doesn't follow the same semantics as a `Try`.
If is peek doesn't succeed, it maintains it doesn't overwrite the value currently contained in the *out* parameter. This behaviour is intentional.  

@gafter  Let me just re run the tests as it looks like the proposal has the wrong dll. The one without `Peep`?

**Comment by AdamSpeight2008** [comment]
@gafter  After rerun the test, with the correct DLL (with Peep) for the Proposal
```
Run #1  00:11:56.5150290
Run #2  00:10:02.4786336
Run #3  00:12:08.7661797
        ----------------
        00:34:07.7598426 sum
        ----------------
        00:11:35.9199475 avg
```
[Zipped Files](http://s000.tinyupload.com/index.php?file_id=06501630485930023407)

**Comment by gafter** [comment]
@AdamSpeight2008 Can you please give an example of where, in your code, you depend on `Peep` not modifying the ref parameter when it returns `False`? I would expect it to be more clear to use the "try" pattern and just use separate variables, no?

**Comment by AdamSpeight2008** [comment]
@gafter  
Any place where is use in the form  where it is the First argument in a short-circuiting condition.
```vb
' ch as some existing value before here
While Peep(Here, ch) AndAlso IsFoo(ch)
...
End While 
Dim bar = Foo (ch)
```
Imagine the While is false because of the `Peep` on the first iteration. The `Foo(ch)` would potentially the wrong value.

[Source](https://github.com/AdamSpeight2008/roslyn-AdamSpeight2008/blob/Scanner(Peep_2)/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L550)
```vb,net
        Private Function ScanLineContinuation(tList As SyntaxListBuilder) As Boolean
            Dim ch As Char
            If Not Peep(0, ch) OrElse
               Not IsAfterWhitespace() OrElse
               Not IsUnderscore(ch) Then
                Return False
            End If
```
Let say `Peep(ch,0)` returns True.
```
            Dim Here = 1
            While Peep(Here, ch) AndAlso IsWhitespace(ch)
                Here += 1
            End While
```
This one `Peep( 1, ch )` return false
```
            ' Line continuation is valid at the end of the
            ' line or at the end of file only.
            Dim atNewLine = IsNewLine(ch)
```
`IsNewLine(ch)` potentially passed the wrong value.

**Comment by AdamSpeight2008** [comment]
[Another Example](https://github.com/AdamSpeight2008/roslyn-AdamSpeight2008/blob/Scanner(Peep_2)/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L2237)  Possible Assert Failure

**Comment by gafter** [comment]
That is way, way too subtle. I'm afraid this code is far too delicate if that is your intent. It is likely to break when next it is maintained. I'd prefer to see that code changed from `IsNewLine(ch)` to `Peek(Here, ch) AndAlso IsNewLine(ch)`. In that case the move to the "try" pattern would be possible.

By the way, local variables should start with lower-case letters; `here` instead of `Here`.

**Comment by AdamSpeight2008** [comment]
@gafter Don't blame me for casing, I just working with what already existed.

**Comment by AdamSpeight2008** [comment]
@gafter
  >That is way, too subtle.    

It maybe so but they are pre-existing within the scanner source, the PR doesn't change that.  It just exposes it because of the PR touch that area of the source.

> At least in the original code this was all explicit.     

It still is explicit, just that `Peep` can exploit the fact that `CanGet` and `Peep` are used in closed proximity, so there is (as the test show) potential for efficiencies. The original code also involved re-lookingup things that where previously used. `Peep` allows us to cache that, whilst not having separate it out into a separate section of code. 

> in cases where the try pattern works, and leave the code the old way where it does not?    

`CanGet( Here ) AndAlso Peek(here) = foo` is a try pattern.

All of the usage can be replace with a `Peep`, I reverted some back to original forms, as the code was simpler.

**Comment by gafter** [comment]
I think the code would be much more clear if this kind of method were only used when the "try" pattern were applicable. A method that takes a ref parameter and sometimes assigns it and sometimes does not, and a caller that takes advantage of the difference, is just way too subtle. That was not a pre-existing pattern in the code. That would also enable you to rename the method so it follows a familar "try" coding pattern.

**Comment by jaredpar** [comment]
> That is way, way too subtle

Completely agree.  I've looked at this PR a few times and never noticed that.

**Comment by AdamSpeight2008** [comment]
@jaredpar and @gafter 
Reexamine ScanLineContinuation in the existing master, you will see this subtle behavhiour in that method. 

First one succeeds, get to While. It fails on CanGet (1st pass). What do the potential two usages of *ch* expect to see?

**Comment by gafter** [comment]
The original code was clear because the assignments were explicit, under conditional statements.

**Comment by AdamSpeight2008** [comment]
It still is explicit under a conditional statement, only that they've been combined into a single function.
The function maintains existing semantics of the original code.

@Pilchie  Any comment?.

**Comment by gafter** [comment]
Though we appreciate your efforts, we do not want to accept this change as written, though we might consider a change that uses the try pattern where appropriate.

To be clear...

> `CanGet( Here ) AndAlso Peek(here) = foo` is a try pattern.

No, that is not a [try pattern](https://startbigthinksmall.wordpress.com/2011/05/12/the-trycan-pattern/). That is code that could be rewritten into the [try pattern](http://csharp.2000things.com/2014/09/29/1192-following-the-tryparse-pattern/). A [try pattern](http://blog.the-dargans.co.uk/2007/01/try-get-design-pattern.html) is a method (in C# syntax)
`bool TrySomething(..., out SomeType result) ...`
used in an idiomatic way
`if (TrySomething(..., out result)) { /* code using result */ }`

The reasons we do not want to pull these changes are
- The code does not fix any bug in Roslyn
- The code does not address any identified performance issues
- The code uses non-idiomatic coding patterns that are likely to result in support issues later
- This is one of the most delicate areas in the compiler, and we are reluctant to change it unless we have a very good reason to do so.

If you're interested in contributing to the Roslyn compilers, I recommend you select issues that we've already identified as [things we'd like fixed and would welcome community contributions](https://github.com/dotnet/roslyn/issues?q=is%3Aopen+is%3Aissue+label%3AArea-Compilers+label%3A%22Up+for+Grabs%22). For example, https://github.com/dotnet/roslyn/issues/2150 is low-hanging fruit.

**Comment by AdamSpeight2008** [comment]
It would perform worse if it followed the other try patterns, as it would involve an extra assignment on each false.

**Comment by AdamSpeight2008** [comment]
There are functions with the current source that have the "Try" prefix,return a boolean and don't have an `<out>` parameter on the pass-back parameter,

 [Block of Four in Parser.vb](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L2405)

**Pull request closed without merging**